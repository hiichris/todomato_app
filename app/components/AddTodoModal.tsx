import React, { useEffect, useState, useRef } from "react";
import {
  Modal,
  View,
  Text,
  Alert,
  Pressable,
  StyleSheet,
  Dimensions,
  TextInput,
  ScrollView,
  KeyboardAvoidingView,
  Platform,
  TouchableWithoutFeedback,
  Keyboard,
} from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";
import { primaryColor } from "../helpers/constants";
import { addNewTodo, getCategories, getTodoById } from "../services/db_service";

// Add Todo Modal Component
export default function AddTodoModal({
  modalVisible,
  setModalVisible,
  setTodos,
  refreshTodos,
  router,
  categories,
  setCategories,
  isPassiveAssignment,
}) {
  const [todoTitle, setTodoTitle] = React.useState("");
  const [selectedCategory, setSelectedCategory] = useState(null);
  const scrollViewRef = useRef(null);

  useEffect(() => {
    // Add a listener to the keyboard
    const keyboardDidShowListener = Keyboard.addListener(
      "keyboardDidShow",
      () => {
        // Scroll to the bottom of the ScrollView when the keyboard is shown
        scrollViewRef.current?.scrollToEnd({ animated: true });
      }
    );

    return () => {
      keyboardDidShowListener.remove();
    };
  }, []);

  useEffect(() => {
    // Retrieve latest categories from the database when the modal is visible
    retrieveDBCategories();
  }, [modalVisible]);

  // Create a new todo
  const createTodoHandler = async () => {
    let lastCreatedTodoId = null;
    // Sanity check
    if (todoTitle === "") {
      Alert.alert("Todo Title is required");
      return;
    }
    if (selectedCategory === null) {
      Alert.alert("Please select a category");
      return;
    }

    console.log(
      "Todo Title: ",
      todoTitle,
      "selectedCategory: ",
      selectedCategory
    );
    // Add the new todo to the database
    addNewTodo(setTodos, todoTitle, selectedCategory)
      .then((result) => {
        console.log("result: ", result);
        lastCreatedTodoId = result.lastInsertRowId;
        refreshTodos();

        // Clear the todoTitle and selectedCategory
        setTodoTitle("");
        setSelectedCategory(null);
        setModalVisible(!modalVisible);

        // Get the last created todo details and navigate to the details screen
        getTodoById(lastCreatedTodoId)
          .then((todo) => {
            router.push({
              pathname: "/todo_details",
              params: {
                passiveTask: isPassiveAssignment,
                title: todo?.title,
                id: todo?.id,
                categoryName: todo?.category_name,
                categoryColor: todo?.category_color,
                refreshTodos: refreshTodos,
              },
            });
          })
          .catch((error) => {
            console.log("Error: ", error);
          });
      })
      .catch((error) => {
        console.log("Error: ", error);
      });

    // Clear the todoTitle
    setTodoTitle("");
    setSelectedCategory(null);
    setModalVisible(!modalVisible);
  };

  // Retrieve categories from the database
  const retrieveDBCategories = async () => {
    const dbCategories = await getCategories(false)
      .then((result) => {
        setCategories(result);
      })
      .catch((error) => {
        console.log("Error: ", error);
      });
  };

  // Handle the onFocus event
  const onFocusHandler = () => {
    if (selectedCategory === null) {
      Alert.alert("Please select a category first!");
      Keyboard.dismiss();
      return;
    }
  };

  return (
    <SafeAreaView>
      <Modal
        animationType="slide"
        transparent={true}
        visible={modalVisible}
        onRequestClose={() => {
          Alert.alert("Modal has been closed.");
          setModalVisible(!modalVisible);
        }}
      >
        <KeyboardAvoidingView
          behavior={Platform.OS === "ios" ? "padding" : "height"}
          style={styles.centeredView}
        >
          <ScrollView
            style={styles.centeredView}
            ref={scrollViewRef}
            keyboardShouldPersistTaps="always"
          >
            <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
              <View style={styles.modalView}>
                <View style={styles.modalHeaderContainer}>
                  <Text style={styles.headerText}>Create Todo</Text>
                  <Pressable
                    style={styles.closeCirleButton}
                    onPress={() => setModalVisible(!modalVisible)}
                  >
                    <Text style={styles.closeCircleText}>X</Text>
                  </Pressable>
                </View>
                <View style={styles.modalBodyContainer}>
                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Category:</Text>
                    <View style={styles.inputGroupContainer}>
                      <ScrollView
                        style={styles.categoryScrollView}
                        horizontal={true}
                        showsHorizontalScrollIndicator={false}
                      >
                        <View style={styles.categoryContainer}>
                          {categories &&
                            categories.map((category) => (
                              <Pressable
                                key={category.id}
                                style={({ pressed }) => [
                                  styles.categoryButton,
                                  {
                                    backgroundColor: pressed
                                      ? "lightgray"
                                      : "white",
                                  },
                                  {
                                    borderColor:
                                      selectedCategory === category.id
                                        ? "#B2361B"
                                        : "white",
                                  },
                                ]}
                                onPress={() => setSelectedCategory(category.id)}
                              >
                                <Text style={styles.categoryButtonText}>
                                  {category.name}
                                </Text>
                              </Pressable>
                            ))}
                        </View>
                      </ScrollView>
                    </View>
                    <View style={styles.spacer}></View>
                  </View>

                  <View style={styles.inputContiner}>
                    <Text style={styles.modalText}>Todo Title:</Text>
                    <TextInput
                      style={styles.modalTextInput}
                      onChangeText={setTodoTitle}
                      onFocus={onFocusHandler}
                      value={todoTitle}
                      returnKeyType="create"
                      numberOfLines={2}
                      multiline={true}
                      textAlign="left"
                      textAlignVertical="top"
                      onSubmitEditing={createTodoHandler}
                    />
                  </View>

                  <View style={styles.buttonsContainer}>
                    <Pressable
                      style={({ pressed }) => [
                        styles.button,
                        styles.buttonCreate,
                        { backgroundColor: pressed ? "#9a2800" : "#B2361B" },
                      ]}
                      onPress={createTodoHandler}
                    >
                      <Text style={styles.createButtonText}>Create</Text>
                    </Pressable>
                  </View>
                </View>
                <View style={{ flexGrow: 2 }}></View>
              </View>
            </TouchableWithoutFeedback>
          </ScrollView>
        </KeyboardAvoidingView>
      </Modal>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  centeredView: {
    flex: 1,
  },
  modalView: {
    flex: 1,
    width: Dimensions.get("window").width,
    height: Dimensions.get("window").height / 2,
    marginTop: Dimensions.get("window").height / 2,
    backgroundColor: "rgba(255, 87, 51, 0.95)",
    borderRadius: 20,
    padding: 16,
    paddingTop: 40,
    alignItems: "center",
    shadowColor: "#000",
    shadowOpacity: 0.15,
    shadowRadius: 5,
    elevation: 5,
    justifyContent: "space-evenly",
  },
  buttonsContainer: {
    margin: 8,
    flexDirection: "row",
    justifyContent: "space-between",
    marginTop: 16,
  },
  button: {
    borderRadius: 20,
    padding: 16,
    paddingHorizontal: 12,
    margin: 8,
    elevation: 2,
  },
  buttonCreate: {
    backgroundColor: "#B2361B",
    borderColor: primaryColor,
    width: "100%",
  },
  createButtonText: {
    color: "white",
    fontWeight: "bold",
    textAlign: "center",
    fontSize: 16,
  },
  modalText: {
    flex: 2,
    textAlign: "center",
    color: "white",
    fontSize: 14,
    fontWeight: "bold",
  },
  modalTextInput: {
    flex: 4,
    height: 40,
    width: Dimensions.get("window").width - 32,
    borderColor: "white",
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    margin: 8,
    color: "white",
    height: 60,
    verticalAlign: "top",
    justifyContent: "flex-start",
  },
  inputContiner: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  modalHeaderContainer: {
    flex: 2,
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  modalBodyContainer: {
    flex: 50,
    justifyContent: "flex-start",
    alignItems: "center",
  },
  headerText: {
    color: "white",
    fontSize: 18,
    fontWeight: "bold",
    marginTop: -38,
  },
  closeCirleButton: {
    marginTop: -38,
    position: "absolute",
    left: Dimensions.get("window").width / 2,
    top: 16,
    backgroundColor: "white",
    borderRadius: 20,
    width: 30,
    height: 30,
    justifyContent: "center",
    alignItems: "center",
  },
  closeCircleText: {
    color: primaryColor,
    fontWeight: "bold",
    fontSize: 10,
    width: 13,
    textAlign: "center",
  },
  inputGroupContainer: {
    flex: 4,
    flexDirection: "row",
    justifyContent: "space-evenly",
  },
  categoryContainer: {
    padding: 0,
    flexDirection: "row",
    height: 50,
    width: "100%",
  },
  categoryButton: {
    flex: 1,
    margin: 4,
    backgroundColor: "white",
    padding: 12,
    paddingVertical: 0,
    borderRadius: 20,
    justifyContent: "center",
    borderWidth: 4,
  },
  categoryButtonText: {
    color: primaryColor,
    fontSize: 14,
    textAlign: "center",
  },
  categoryScrollView: {
    marginRight: 8,
  },
  spacer: {
    height: 100,
  },
});
