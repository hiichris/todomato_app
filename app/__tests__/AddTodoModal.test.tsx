import React from "react";
import renderer from "react-test-renderer";
import { waitFor } from "@testing-library/react-native";

import AddTodoModal from "../components/AddTodoModal";


describe("AddTodoModal", () => {
  it("renders correctly", async () => {
    const tree = renderer.create(<AddTodoModal />).toJSON();
    await waitFor(() => expect(tree).toMatchSnapshot());
  });
});
