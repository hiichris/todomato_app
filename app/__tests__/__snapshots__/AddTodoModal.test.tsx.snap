// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodoModal renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={true}
  >
    <View
      onLayout={[Function]}
      style={
        [
          {
            "flex": 1,
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <RCTScrollView
        keyboardShouldPersistTaps="always"
        style={
          {
            "flex": 1,
          }
        }
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgba(255, 87, 51, 0.95)",
                "borderRadius": 20,
                "elevation": 5,
                "flex": 1,
                "height": 667,
                "justifyContent": "space-evenly",
                "marginTop": 667,
                "padding": 16,
                "paddingTop": 40,
                "shadowColor": "#000",
                "shadowOpacity": 0.15,
                "shadowRadius": 5,
                "width": 750,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 2,
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 18,
                    "fontWeight": "bold",
                    "marginTop": -38,
                  }
                }
              >
                Create Todo
              </Text>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "white",
                    "borderRadius": 20,
                    "height": 30,
                    "justifyContent": "center",
                    "left": 375,
                    "marginTop": -38,
                    "position": "absolute",
                    "top": 16,
                    "width": 30,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#FF5733",
                      "fontSize": 10,
                      "fontWeight": "bold",
                      "textAlign": "center",
                      "width": 13,
                    }
                  }
                >
                  X
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 50,
                  "justifyContent": "flex-start",
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "white",
                      "flex": 2,
                      "fontSize": 14,
                      "fontWeight": "bold",
                      "textAlign": "center",
                    }
                  }
                >
                  Category:
                </Text>
                <View
                  style={
                    {
                      "flex": 4,
                      "flexDirection": "row",
                      "justifyContent": "space-evenly",
                    }
                  }
                >
                  <RCTScrollView
                    horizontal={true}
                    showsHorizontalScrollIndicator={false}
                    style={
                      {
                        "marginRight": 8,
                      }
                    }
                  >
                    <View>
                      <View
                        style={
                          {
                            "flexDirection": "row",
                            "height": 50,
                            "padding": 0,
                            "width": "100%",
                          }
                        }
                      />
                    </View>
                  </RCTScrollView>
                </View>
                <View
                  style={
                    {
                      "height": 100,
                    }
                  }
                />
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "white",
                      "flex": 2,
                      "fontSize": 14,
                      "fontWeight": "bold",
                      "textAlign": "center",
                    }
                  }
                >
                  Todo Title:
                </Text>
                <TextInput
                  multiline={true}
                  numberOfLines={2}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  onSubmitEditing={[Function]}
                  returnKeyType="create"
                  style={
                    {
                      "borderColor": "white",
                      "borderRadius": 10,
                      "borderWidth": 1,
                      "color": "white",
                      "flex": 4,
                      "height": 60,
                      "justifyContent": "flex-start",
                      "margin": 8,
                      "padding": 10,
                      "verticalAlign": "top",
                      "width": 718,
                    }
                  }
                  textAlign="left"
                  textAlignVertical="top"
                  value=""
                />
              </View>
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "margin": 8,
                    "marginTop": 16,
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "borderRadius": 20,
                        "elevation": 2,
                        "margin": 8,
                        "padding": 16,
                        "paddingHorizontal": 12,
                      },
                      {
                        "backgroundColor": "#B2361B",
                        "borderColor": "#FF5733",
                        "width": "100%",
                      },
                      {
                        "backgroundColor": "#B2361B",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      {
                        "color": "white",
                        "fontSize": 16,
                        "fontWeight": "bold",
                        "textAlign": "center",
                      }
                    }
                  >
                    Create
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "flexGrow": 2,
                }
              }
            />
          </View>
        </View>
      </RCTScrollView>
    </View>
  </Modal>
</RNCSafeAreaView>
`;
